rfcs/0667-apikey-delegated-permissions.md

```diff
@@ -0,0 +1,124 @@
+# API key delegated permissions
+
+Instead of maintaining separate principal types and access rules for regular users vs API keys, switch to the model whereby a user creates API keys and delegates one or more of the user's permissions to it.
+
+## Motivation
+
+The [current API key system](https://docs.getseq.net/docs/api-keys) was designed for ingestion management only, with limited read capability added on later, and shows a lot of quirks because of it.
+
+* API key tokens are treated primarily as identifiers rather than as credentials
+* It's not possible to create an API key that can perform admin-level tasks at all
+* API keys can only be assigned all-or-nothing user-level access, where more restricted access may be desirable
+* Users can't create API keys to perform tasks on their behalf that they're otherwise permitted to perform themselves
+
+In Seq 5 we have the opportunity to modernize and redesign API keys, and to overhaul the underlying permissions model to better support them.
+
+## Detailed design
+
+The proposal has three major parts:
+
+ * User _roles and permissions_ will be introduced so that API keys can have fine-grained user permissions delegated to them
+ * API keys will gain the concept of an owner - the user who delegates their permissions to the API key
+ * API key storage will switch to a hashed-credentials model
+
+### Usage
+
+The summaries in this section highlight how the major use cases for API keys will be achieved in the new version.
+
+#### API keys for application log ingestion
+
+It's anticipated that API keys for application log ingestion will continue to be created and maintained by a system administrator, as shared (non-personal) API keys. This will prevent interruption of application logging when users come and go.
+
+A user may still create a personal API key and use it to ingest log events if desired - this essentially provides "self-service" API key creation. Using shared keys for ingesion is encouraged by recommendation only.
```

Comment:
What do we mean by _encouraged by recommendation_ here?

Reply:
Will clarify - I meant to say that we'll recommend it but not take any steps to force it (it's a bit of a discretionary point).